File: app/src/main/cpp/wn-steam-client/src/pb/cfamilygroups.cpp
#include "wn_steam/pb/cfamilygroups.h"

#include "wn_steam/proto_wire.h"

namespace wn_steam::pb {

std::vector<uint8_t> CFamilyGroups_GetFamilyGroup_Request::serialize() const {
    std::vector<uint8_t> out;
    proto::Writer w(out);
    w.uint64_field(1, family_groupid);
    return out;
}

std::optional<FamilyGroupMember>
FamilyGroupMember::deserialize(std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    FamilyGroupMember m;
    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        if (t->field_number == 1) {
            if (auto v = r.fixed64(); v) m.steamid = *v;
            else return std::nullopt;
        } else if (!r.skip(t->wire_type)) {
            return std::nullopt;
        }
    }
    return m;
}

std::optional<CFamilyGroups_GetFamilyGroup_Response>
CFamilyGroups_GetFamilyGroup_Response::deserialize(
        std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    CFamilyGroups_GetFamilyGroup_Response m;
    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        switch (t->field_number) {
            case 1:
                if (auto v = r.string(); v) m.name = std::move(*v);
                else return std::nullopt;
                break;
            case 2: {
                auto b = r.bytes();
                if (!b) return std::nullopt;
                auto member = FamilyGroupMember::deserialize(*b);
                if (!member) return std::nullopt;
                m.members.push_back(std::move(*member));
                break;
            }
            default:
                if (!r.skip(t->wire_type)) return std::nullopt;
                break;
        }
    }
    return m;
}

}  // namespace wn_steam::pb

File: app/src/main/cpp/wn-steam-client/rust/src/pb/cfamilygroups.rs
use crate::proto_wire::{Reader, Writer};

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct CFamilyGroupsGetFamilyGroupRequest {
    pub family_groupid: u64,
}

impl CFamilyGroupsGetFamilyGroupRequest {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        Writer::new(&mut out).uint64_field(1, self.family_groupid);
        out
    }
}

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct FamilyGroupMember {
    pub steamid: u64,
}

impl FamilyGroupMember {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                1 => msg.steamid = reader.fixed64()?,
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct CFamilyGroupsGetFamilyGroupResponse {
    pub name: String,
    pub members: Vec<FamilyGroupMember>,
}

impl CFamilyGroupsGetFamilyGroupResponse {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                1 => msg.name = reader.string()?,
                2 => msg
                    .members
                    .push(FamilyGroupMember::deserialize(reader.bytes()?)?),
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_family_group_members() {
        let mut member = Vec::new();
        Writer::new(&mut member).fixed64_field(1, 123);

        let mut body = Vec::new();
        {
            let mut w = Writer::new(&mut body);
            w.string_field(1, "Family");
            w.submessage_field(2, &member);
        }

        let parsed = CFamilyGroupsGetFamilyGroupResponse::deserialize(&body).unwrap();
        assert_eq!(parsed.name, "Family");
        assert_eq!(parsed.members[0].steamid, 123);
    }
}
